src/mpid/ch4/netmod/ofi/ofi_win.c

```diff
@@ -192,7 +192,7 @@ static int win_allgather(MPIR_Win * win, void *base, int disp_unit)
     /* Don't register MR for NULL buffer, because FI_MR_BASIC mode requires
      * that all registered memory regions must be backed by physical memory
      * pages at the time the registration call is made. */
-    if ((!MPIDI_OFI_ENABLE_MR_PROV_KEY && !MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS) || base) {
+    if ((!MPIDI_OFI_ENABLE_MR_PROV_KEY && !MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS) && base) {
```

Comment:
The change here seems incorrect. The original code means registering MR if `scalable MR || base!= NULL`, but the changed code means registering only when it is `scalable MR && base!=NULL`. I think what we want is to register MR only when `base!=NULL`, no matter is scalable MR or basic MR.

Reply:
If `MPIDI_OFI_ENABLE_MR_PROV_KEY` is enabled, it means the provider does not need additional `mr_register`, right? We probably need set `mr` to the provider key here instead of `mr_register`.

`MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS` -- this one I am not  sure, does it mean that, if *not* enabled, then we should register direct pointer address?

So I think neither the original logic nor the new logic is exactly correct. We may need consider each case (and some combinations of them) separately.

Reply:
After checking with documentation, I agree with you that we probably can skip the checking of capabilities here. I suspect our other places checking MPIDI_OFI_ENABLE_MR_PROV_KEY and MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS are not exactly correct ... I'll follow up with  it.

Reply:
No, I think `MPIDI_OFI_ENABLE_MR_PROV_KEY ` have different meaning. See here: https://ofiwg.github.io/libfabric/v1.10.1/man/fi_mr.3.html

Reply:
I actually don't see why we need check these capabilities when registering MR if we never do OFI RMA for dynamic window. I believe all OFI RMA calls require a call to `fi_mr_reg` to get the `mr_key`.

I think the only reason why the capabilities check exists was that, for provider that does not require `FI_MR_VIRT_ADDR` (no need to ping to physical address), the base can be `NULL`. Although there is no documentation for it.

@zhenggb72 Are you also developing dynamic window? This PR might be relevant.

Reply:
Sorry, I was on vacation last week.  I agree, we don't need to check those capabilities if we always fallback for dynamic window. Alternatively, we can always try to call the fi_mr_reg. The problem was that MPIDI_OFI_CALL which returns error, instead, we can choose to ignore the error in case it occurs and leave MPIDI_OFI_WIN(win).mr as NULL.

Reply:
>  ...instead, we can choose to ignore the error in case it occurs and leave MPIDI_OFI_WIN(win).mr as NULL.

I think I like that idea if it is robust. Is it possible that a provider will return a non-NULL `mr` as we trying to register a `NULL` address and later on give us error when we try to use it?

Reply:
I think we should only check for the return value from fi_mr_reg, if it not 0, set mr to NULL to make sure it does not accidentally get a garbage value. 

If we want to do this, I think we should keep the checking for those capabilities like it was, because we don't even bother trying fi_mr_reg if it is not scalable MR mode with base = NULL. So it will be like calling fi_mr_reg when base != NULL, or  base == NULL, but scalable_mr is on.

Reply:
@zhenggb72 I think the major reason why we want to get rid of the capabilities check in this PR is that, we are unclear what the behavior would be if we assign a NULL base, no matter it is scalable MR (now with some combination of capabilities) or basic MR. In other words, there is no clear OFI documentation for it, so behavior may be undefined for some providers. For example:
- If it fails, it may return error for the fi_mr_reg call, but it may also segfault at a later arbitrary call.
- If it succeeds, what does it mean? Dose it mean that the entire virtual address has been registered and any virtual address can be used at a later fi_read/write call?

If the OFI manual adds the definition of the behavior of fi_mr_reg with NULL base, we can add the optimization back into MPICH at that time, as you suggested. What do you think?

Besides, I would add that the optimization is not critical as it helps only a very special case (i.e., dynamic window with scalable MR).

Reply:
ok, sounds good to me.